xds/internal/balancer/priority/balancer.go

```diff
@@ -0,0 +1,214 @@
+	newConfig, ok := s.BalancerConfig.(*lbConfig)
+	if !ok {
+		return fmt.Errorf("unexpected balancer config with type: %T", s.BalancerConfig)
+	}
+	addressesSplit := hierarchy.Group(s.ResolverState.Addresses)
+
+	pb.mu.Lock()
+	defer pb.mu.Unlock()
+	// Create and remove children, since we know all children from the config
+	// are used by some priority.
+	for name, newSubConfig := range newConfig.Children {
+		bb := balancer.Get(newSubConfig.Config.Name)
+		if bb == nil {
+			pb.logger.Warningf("balancer name %v from config is not registered", newSubConfig.Config.Name)
+			continue
+		}
+
+		currentChild, ok := pb.children[name]
+		if !ok {
+			// This is a new child, add it to the children list. But note that
+			// the balancer isn't built, because this child can be a low
+			// priority. If necessary, it will be built when syncing priorities.
+			cb := newChildBalancer(name, pb, bb)
+			cb.updateConfig(newSubConfig.Config.Config, resolver.State{
+				Addresses:     addressesSplit[name],
+				ServiceConfig: s.ResolverState.ServiceConfig,
+				Attributes:    s.ResolverState.Attributes,
+			})
+			pb.children[name] = cb
+			continue
+		}
+
+		// This is not a new child. But the config/addresses could change.
+
+		// The balancing policy name is changed, close the old child. But don't
+		// rebuild, rebuild will happen when syncing priorities.
+		if currentChild.bb.Name() != bb.Name() {
+			currentChild.stop()
+			currentChild.bb = bb
+		}
+
+		// Update config and address, but note that this doesn't send the
+		// updates to child balancer (the child balancer might not be built, if
+		// it's a low priority).
+		currentChild.updateConfig(newSubConfig.Config.Config, resolver.State{
+			Addresses:     addressesSplit[name],
+			ServiceConfig: s.ResolverState.ServiceConfig,
+			Attributes:    s.ResolverState.Attributes,
+		})
+	}
+
+	// Remove child from children if it's not in new config.
+	for name, oldChild := range pb.children {
+		if _, ok := newConfig.Children[name]; !ok {
+			oldChild.stop()
+		}
+	}
+
+	// Update priorities and handle priority changes.
+	pb.priorities = newConfig.Priorities
+	pb.childToPriority = make(map[string]int)
+	for pi, pName := range newConfig.Priorities {
+		pb.childToPriority[pName] = pi
+	}
+	// Sync the states of all children to the new updated priorities. This
+	// include starting/stopping child balancers when necessary.
+	pb.syncPriority()
+
+	return nil
+}
+
+func (pb *priorityBalancer) ResolverError(err error) {
+	pb.bg.ResolverError(err)
+}
+
+func (pb *priorityBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	pb.bg.UpdateSubConnState(sc, state)
+}
+
+func (pb *priorityBalancer) Close() {
+	pb.done.Fire()
+	pb.bg.Close()
+
+	pb.mu.Lock()
+	defer pb.mu.Unlock()
+	// Clear states of the current child in use, so if there's a race in picker
+	// update, it will be dropped.
+	pb.childInUse = ""
+	if timer := pb.priorityInitTimer; timer != nil {
```

Comment:
What if we make sure `pb.priorityInitTimer` is always non-nil and valid?  Then this can be simply `pb.priorityInitTimer.Stop()` (1 line vs. 4).

Or make a function for this and keep the `nil`: `pb.stopPriorityInitTimer()`.  This pattern appears several times.